packages/flutter/lib/src/material/snack_bar.dart

```diff
@@ -132,24 +138,106 @@ class _SnackBarActionState extends State<SnackBarAction> {
 
   @override
   Widget build(BuildContext context) {
-    Color? resolveForegroundColor(Set<MaterialState> states) {
-      final SnackBarThemeData snackBarTheme = Theme.of(context).snackBarTheme;
-      if (states.contains(MaterialState.disabled)) {
-        return widget.disabledTextColor ?? snackBarTheme.disabledActionTextColor;
+    final SnackBarThemeData defaults = Theme.of(context).useMaterial3
+        ? _SnackbarDefaultsM3(context)
+        : _SnackbarDefaultsM2(context);
+    final SnackBarThemeData snackBarTheme = Theme.of(context).snackBarTheme;
+
+    MaterialStateColor resolveForegroundColor() {
+      if (widget.textColor is MaterialStateColor) {
+        return widget.textColor! as MaterialStateColor;
       }
-      return widget.textColor ?? snackBarTheme.actionTextColor;
+      return MaterialStateColor.resolveWith((Set<MaterialState> states) {
+        if (states.contains(MaterialState.disabled)) {
+          return widget.disabledTextColor ??
+              snackBarTheme.disabledActionTextColor ??
+              defaults.disabledActionTextColor!;
+        }
+        return widget.textColor ??
+            snackBarTheme.actionTextColor ??
+            defaults.actionTextColor!;
+      });
     }
 
     return TextButton(
       style: ButtonStyle(
-        foregroundColor: MaterialStateProperty.resolveWith<Color?>(resolveForegroundColor),
+        foregroundColor: resolveForegroundColor(),
       ),
       onPressed: _haveTriggeredAction ? null : _handlePressed,
       child: Text(widget.label),
     );
   }
 }
 
+/// An icon for a [SnackBar].
+///
+/// The snack bar icon is not included by default. To enable it, include the
+/// [icon] parameter in the snack bar.
+///
+/// By default, snack bar icons use the Close icon.
+///
+/// Snack bar icons can only be pressed once. Subsequent presses are ignored.
+/// By default, tapping the icon will close the snackbar.
+///
+/// See also:
+///
+///  * [SnackBar]
+///  * <https://material.io/design/components/snackbars.html>
+class SnackBarIcon extends StatefulWidget {
```

Comment:
Does this need to be public? Why do we need a whole new widget for the close icon?

Reply:
I don't think the icon is meant to be interactive, it is?

Reply:
In the spec it's mentioned as a 'Close Icon' but doesn't specify interactivity. Would you propose making it non-customizable, and only have the closing functionality available?

Reply:
> Would you propose making it non-customizable, and only have the closing functionality available?

Yes, maybe a `showCloseIcon` flag or similar.

Reply:
Sure - that simplifies things. Removed the SnackBarIcon class, and updated the tests accordingly.